tests/client/net_handler_client.c

```diff
@@ -0,0 +1,520 @@
+
+void start_net_handler ()
+{
+
+	//fork net_handler process
+	if ((nh_pid = fork()) < 0) {
+		printf("fork: %s\n", strerror(errno));
+	} else if (nh_pid == 0) { //child
+		//cd to the net_handler directory
+		if (chdir("../../net_handler") == -1) {
+			printf("chdir: %s\n", strerror(errno));
+		}
+
+		//exec the actual net_handler process
+		if (execlp("./net_handler", "net_handler", (char *) 0) < 0) {
+			printf("execlp: %s\n", strerror(errno));
+		}
+	} else { //parent
+		sleep(1); //allows net_handler to set itself up
+
+		//Connect to the raspi networking ports to catch network output
+		nh_tcp_dawn_fd = connect_tcp(DAWN_CLIENT);
+		nh_tcp_shep_fd = connect_tcp(SHEPHERD_CLIENT);
+		if ((nh_udp_fd = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
+			printf("socket: UDP socket creation failed...\n");
+			stop_net_handler();
+			exit(1);
+		}
+		memset(&udp_servaddr, 0, sizeof(struct sockaddr_in));
+		udp_servaddr.sin_family = AF_INET; 
+		udp_servaddr.sin_addr.s_addr = inet_addr(RASPI_ADDR); 
+		udp_servaddr.sin_port = htons(RASPI_UDP_PORT); 
+		
+		//open /dev/null
+		null_fp = fopen("/dev/null", "w");
+
+		//init the mutex that will control whether udp prints to screen
+		if (pthread_mutex_init(&print_udp_mutex, NULL) != 0) {
+			printf("pthread_mutex_init: print udp mutex\n");
+		}
+		print_next_udp = 0;
+		udp_output_fp = null_fp; //by default set to output to /dev/null
+
+		//start the thread that is dumping output from net_handler to stdout of this process
+		if (pthread_create(&dump_tid, NULL, output_dump, NULL) != 0) {
+			printf("pthread_create: output dump\n");
+		}
+		sleep(1); //allow time for thread to dump output before returning to client
+	}
+}
+
+void stop_net_handler ()
+{
+	//send signal to net_handler and wait for termination
+	if (kill(nh_pid, SIGINT) < 0) {
+		printf("kill: %s\n", strerror(errno));
+	}
+	if (waitpid(nh_pid, NULL, 0) < 0) {
+		printf("waitpid: %s\n", strerror(errno));
+	}
+	
+	//killing net handler should cause dump thread to return, so join with it
+	if (pthread_join(dump_tid, NULL) != 0) {
+		printf("pthread_join: output dump\n");
+	}
+	
+	//close all the file descriptors
+	if (nh_tcp_shep_fd != -1) {
+		close(nh_tcp_shep_fd);
+	}
+	if (nh_tcp_dawn_fd != -1) {
+		close(nh_tcp_dawn_fd);
+	}
+	if (nh_udp_fd != -1) {
+		close(nh_udp_fd);
+	}
+}	
+	
+
+void send_run_mode (int client, int mode)
+{
+	RunMode run_mode = RUN_MODE__INIT;
+	uint8_t *send_buf;
+	uint16_t len;
+	
+	//set the right mode
+	switch (mode) {
+		case (IDLE_MODE):
+			run_mode.mode = MODE__IDLE;
+			break;
+		case (AUTO_MODE):
+			run_mode.mode = MODE__AUTO;
+			break;
+		case (TELEOP_MODE):
+			run_mode.mode = MODE__TELEOP;
+			break;
+		default:
+			printf("ERROR: sending run mode message\n");
+	}
+	
+	//build the message
+	len = run_mode__get_packed_size(&run_mode);
+	send_buf = make_buf(RUN_MODE_MSG, len);
+	run_mode__pack(&run_mode, send_buf + 3);
```

Comment:
For all these `+ 3`, instead use ` + BUFFER_OFFSET`

Reply:
👌